Framework/PythonInterface/mantid/utils/dgs/_geometry.py

```diff
@@ -0,0 +1,299 @@
+    Returns
+    """
+    pass
+
+
+class ErrorCodes(Enum):
+    """
+    Error codes for qangle
+        - 0 correct result
+        - 1 input parameters are wrong for the corresponding hkl
+        - 2 conservation of energy and momentum can't be satisfied
+        - 3 out of plane angle too high
+        - 4 scattering outside detector
+        - 5 scattering in the beam stop
+        - 6 goniometer out of range
+    """
+    CORRECT = 0
+    WRONG_INPUT = 1
+    QE_CONSERVATION = 2
+    OUT_OF_PLANE = 3
+    OUTSIDE_DETECTOR = 4
+    INSIDE_BEAMSTOP = 5
+    GONIOMETR = 6
+
+
+def _qangle_validate_inputs(hkl: np.array,
+                            Ei: float or np.array,
+                            DeltaE: float or np.array,
+                            sign: float or np.array,
+                            lattice: OrientedLattice,
+                            detector_constraints: bool,
+                            horizontal_extent: np.array,
+                            vertical_extent: np.array,
+                            horizontal_extent_low: np.array,
+                            vertical_extent_low: np.array,
+                            goniometer_constraints: bool,
+                            goniometer_range):
+    """
+    Validate inputs for qangle function, according to the rules for
+    that function
+    """
+    try:
+        len_hkl = len(hkl)
+        if len(hkl[0]) != 3:
+            raise ValueError()
+    except (TypeError, ValueError):
+        raise ValueError('hkl is not an array of triplets')
+
+    try:
+        # check if float
+        Ei = float(Ei)
+        Ei = np.full(len_hkl, Ei)
+    except ValueError:
+        raise ValueError('Ei is not a float or numpy array')
+    except TypeError:
+        if len(Ei) != len_hkl:
+            raise ValueError('Ei has different length than hkl')
+    try:
+        # check if float
+        DeltaE = float(DeltaE)
+        DeltaE = np.full(len_hkl, DeltaE)
+    except ValueError:
+        raise ValueError('DeltaE is not a float or numpy array')
+    except TypeError:
+        if len(DeltaE) != len_hkl:
+            raise ValueError('DeltaE has different length than hkl')
+
+    try:
+        # check if int
+        sign = int(sign)
+        sign = np.full(len_hkl,sign)
+    except ValueError:
+        raise ValueError('sign is not an int or numpy array')
+    except TypeError:
+        if len(sign) != len_hkl:
+            raise ValueError('sign has different length than hkl')
+
+    try:
+        UB = lattice.getUB() * 2. * np.pi
+    except:
+        raise ValueError("Can't get the UB matrix from the lattice object")
+
+    # inputs for geometry and goniometer constraints
+    if detector_constraints:
+        if horizontal_extent[0]<-180 or horizontal_extent[1]<horizontal_extent[0] or horizontal_extent[1]>180:
+            raise ValueError("Horizontal constraints must obey -180 <= horizontal_extent[0] <= "
```

Comment:
This is more of a user feedback issue, could we print out what the user entered in the error message as it makes it much easier for users to find the error in a sea of numbers using find / visually scanning through

```
raise ValueError("Horizontal constraints must obey -180 <= horizontal_extent_low[0] <= "
                             f"horizontal_extent_low[1] <=180. Got {horizontal_extent_low[0]} and {horizontal_extent_low[1]}."
```

Reply:
Good point